desc/compute/utils.py

```diff
@@ -0,0 +1,272 @@
+        in an array that matches the grid's pattern.
+    surface_label : str
+        The surface label of rho, theta, or zeta.
+
+    Returns
+    -------
+    compress_x : ndarray
+        x[grid.unique_surface_label_indices]
+        This array will be sorted such that the
+            first element corresponds to the value associated with the smallest surface
+            last element  corresponds to the value associated with the largest surface
+
+    """
+    assert surface_label in {"rho", "theta", "zeta"}
+    assert len(x) == grid.num_nodes
+    if surface_label == "rho":
+        return x[grid.unique_rho_idx]
+    if surface_label == "theta":
+        return x[grid.unique_theta_idx]
+    if surface_label == "zeta":
+        return x[grid.unique_zeta_idx]
+
+
+def expand(grid, x, surface_label="rho"):
+    """Expand x by duplicating elements to match the grid's pattern.
+
+    Parameters
+    ----------
+    grid : Grid
+        Collocation grid containing the nodes to evaluate at.
+    x : ndarray
+        Stores the values of a surface function (a function constant over a surface)
+        for all unique surfaces of the specified label on the grid.
+        - len(x) should be grid.num_surface_label
+        - x should be sorted such that the
+            first element corresponds to the value associated with the smallest surface
+            last element  corresponds to the value associated with the largest surface
+    surface_label : str
+        The surface label of rho, theta, or zeta.
+
+    Returns
+    -------
+    expand_x : ndarray
+        X expanded to match the grid's pattern.
+
+    """
+    assert surface_label in {"rho", "theta", "zeta"}
+    if surface_label == "rho":
+        assert len(x) == grid.num_rho
+        return x[grid.inverse_rho_idx]
+    if surface_label == "theta":
+        assert len(x) == grid.num_theta
+        return x[grid.inverse_theta_idx]
+    if surface_label == "zeta":
+        assert len(x) == grid.num_zeta
+        return x[grid.inverse_zeta_idx]
+
+
+def surface_integrals(grid, q=1, surface_label="rho", match_grid=False):
+    """Compute the surface integral of a quantity for all surfaces in the grid.
+
+    Parameters
+    ----------
+    grid : Grid
+        Collocation grid containing the nodes to evaluate at.
+    q : ndarray
+        Quantity to integrate.
+    surface_label : str
+        The surface label of rho, theta, or zeta to compute integration over.
+    match_grid : bool
+        Whether to expand the result to match the dimension of the grid.
+        If False (default), the result is a single value for each surface in the grid.
+        If True, the result has repeated values to match the number of grid nodes.
+
+    Returns
+    -------
+    integrals : ndarray
+        Surface integrals of q over each surface in grid.
+
+    """
+    nodes, unique_idx, ds = _get_grid_surface(grid, surface_label)
```

Comment:
Might be helpful to retain the comment I had describing what the bins and histogram are doing.
```
    # Separate collocation nodes into bins with boundaries at unique values of the surface label.
    # This groups nodes with identical surface label values.
    # Each is assigned a weight of their contribution to the integral.
    # The elements of each bin are summed, performing the integration.
```
Also I'm fine with the change, but you can make the max_surface_val = 7 without drawbacks. I remember reading that jax didn't like if statements so just mentioning this.

Reply:
I thought this was relatively self-explanatory code, and there is already good documentation on the `histogram` function provided by NumPy and JAX that we don't need to reproduce. But I guess more comments are never bad if you want to add it back in

I know it was fine, but hard-coding the number `7` felt weird and I was trying to make it more understandable. Could we always use `np.inf` instead?

Reply:
Yes I looked through numpy’s code; an infinite bin length shouldn’t slow it down. Could do something like
`max_surface_val = jnp.inf  # can overshoot`. or `surface_label_supremum` if you wanted to be fancy.

I’ll add the comment back in. It’s easy to understand now having read the comment beforehand. But it’s not obvious it does a non-contiguous element integration. For one thing numpy doesn’t explain what the weights argument does well.  I think it’s good to add comments like this so the person reading the code
- can understand immediately how these functions are used in this context without spending hours going through numpy docs. 
- has confidence the algorithm is not a hack. Understanding the current algorithm may stop an unneeded change later